Add WooCommerce Job Queue using Action Scheduler for Async Webhook Delivery

thenbrent opened the pull request:
This is a prototype to demonstrate what is needed to use Action Scheduler as the event queue for async webhook delivery (#18003), after [Slack discussion on the topic](https://woocommercecommunity.slack.com/archives/C4TNYTR28/p1525847140000049). I thought a PR would be a better channel for discussing and demonstrating it than Slack.

~The actual code is not intended to be merged as is, as the version of Action Scheduler included is a pre-release/dev version. If the use of Action Scheduler is considered to be a suitable solution to #18003, I will submit a different PR with the finalised version of Action Scheduler v2.0.~ *Edit:* PR has been updated.

### Commits of Note

Most of the 11,000 additions here can be ignored, because they are just adding the Action Scheduler library. The commits that matter are:

* SHA: 9101d02c2c4d4e - adds an async webhook delivery to the Action Scheduler queue
* SHA: 3531f26a17 - uses an Action Scheduler feature that isn't available with WP Cron or some other background processes to _group_ webhook deliveries, which allows them to be managed specifically rather than being batched with all other scheduled actions. This is handy on large sites that may want to run a specific webhook queue, and allow other actions to run in other queues.

### Outstanding Question

One difference between Action Scheduler's `wc_schedule_single_action()` and WordPress' `wp_schedule_single_event()` functions is that [with the later](https://developer.wordpress.org/reference/functions/wp_schedule_single_event/#description):

> scheduling an event to occur within 10 minutes of an existing event with the same action hook will be ignored unless you pass unique $args values for each scheduled event.

I believe that means that currently, webhooks scheduled for the same objects for different events that happen within 10 minutes (or more specifically, within the time between when the event occurs, and WP Cron processes the webhook) will not trigger two webhooks. Instead, only one will be triggered.

However, using `wc_schedule_single_action()`, both will be triggered (and the existing hook would need to be manually removed if desired).

For example, if an order's status is changed to on-hold, then processing within a few seconds, and WP Cron has not processed the webhook for the on-hold status change yet, then only the webhook for the processing status change would be triggered when using `wp_schedule_single_event()`. When using `wc_schedule_single_event()`, a webhook for both status transitions would be sent. This is very closely related to https://github.com/woocommerce/woocommerce/issues/15699, and using Action Scheduler with more data in the webhook's `$arg` may provide a solution for it.

### Manual Testing

A good way to test this is to run a process that creates a few thousand webhooks (or even better, 10,000s of webhooks), then watch Action Scheduler process them.

There is an admin UI to show the list and current status of actions, see https://github.com/Prospress/action-scheduler#managing-scheduled-actions

Running them via WP CLI would also be a good idea as a test site likely has WP Cron triggered infrequently, and out-of-the-box, Action Scheduler relies on WP Cron to process the queue. Docs on running a Action Scheduler queue via WP CLI here: https://github.com/Prospress/action-scheduler#wp-cli

Commit:
Load Action Scheduler lib

Commit:
Use Action Scheduler for async webhook delivery

Commit:
Add webhook group for scheduled aysnc deliveries

This will make it possible in future to run queues specifically for
delivering webhooks, instead of deliering webhooks in the same queue
as other actions, like scheduled payments, emails or other actions
scheduled by other plugins. This can be used to provide better
control over webhook delivery, which allows for sending webhooks more
quickly than if they are only included in batches of other actions.

Commit:
Set array keys on scheduled webhook action

To make it clearer in the database what the args refer to, e.g.
instead of storing the scheduled action's args as:

[1,123]

This patch changes it so that the args are stored as:

["webhook_id":1,"arg":123]

This has no impact on callbacks on the hooks, they are still passed
the args as normal.

mikejolley commented:
Thanks @thenbrent. Asked @kloon to take a look.

When tables are used, I'm thinking this would definitely be viable for things like webhooks and email sending, but I have doubts about it being used for sequential queue type things such as the idea to handle stock for new orders in a queue. Will review anyhow.

mikejolley commented:
@thenbrent This looks solid and I want to get some performance testing and general testing kicked off with your v2.0. Looks like webhooks (possibly emails?) would be great use cases for this.

I'm not convinced this will be a solution for the order processing queue since those need to run in a specific order, but that shouldn't block this improvement.

thenbrent commented:
> those need to run in a specific order

We've addressed sequential processing now with https://github.com/Prospress/action-scheduler/pull/139

With AS 2.0 and newer, actions are claimed by time and then ID, so two actions scheduled for the same time would be triggered in the same order they were scheduled.

> This looks solid and I want to get some performance testing and general testing kicked off with your v2.0.

What sort of tests do you want to run?

We mostly test it via Subscriptions (e.g. process 100,000 renewal payments scheduled fo the same time), but I can definitely help with more generic test data.

> This looks solid

Awesome! I've just updated this PR to include the latest `2.0.0-beta-1`.

Among other things, that includes a way to claim by group also. That makes it possible to do things like run webhooks via WP CLI that are dedicated just to webhook processing (and have renewals, email and other queues running concurrent).

I've been working with large stores a lot the last 6 months and found that WP CLI is by far the best way to process large groups of async tasks (much better than the default WP Cron runner obviously, but also other approaches, like those used in `WP_Async_Request` / `WP_Background_Process`).

mikejolley commented:
@thenbrent When do you reckon 2.0 will be finalised?

@kloon @claudiulodro What else needs to happen before this can be considered for inclusion? Would be great to get some performance improvements in :) Might also be viable for emails as mentioned, or the sync for orders/stats.

thenbrent commented:
> @thenbrent When do you reckon 2.0 will be finalised?

About a week.

We [tagged 2.0.0-RC-2](https://github.com/Prospress/action-scheduler/commit/bc82d999d35e3dd1aaa6daf63defbfcae6f0b836) yesterday, and expect that will be the last pre-release version. I'll update this PR once the official 2.0.0 version is tagged.

Does that timeline work for WC?

mikejolley commented:
Yep, 3.5 won't ship until end of q3 earliest

kloon commented:
Since @claudiulodro is currently specking out lookup tables, that might be a good thing to try and implement this on for populating the tables. I can assist him on that.

I have been going through the code the last week and noticed one thing that stood out for me, it seems that every time a batch is run it will flush the whole object cache completely. That might become a bit of a performance issue if you have batches kicking off every minute, for instance say we want to start processing orders via this that will mean we will probably schedule a new batch every minute which will clear the object cache that often.

I have not gone into a deeper investigation but just something I saw while going through the code.

thenbrent commented:
> it seems that every time a batch is run it will flush the whole object cache completely

The object cache is only cleared when **not** using an external object cache, so on large sites, which hopefully are using Redis or similar, the cache wouldn't be cleared.

Correct me if I'm wrong (I'm no expert on it), but I also believe it's pretty low impact to clear the object cache when it's WP's default object cache, because it's just request based. Meaning it just [loads and/or keeps things in memory for the duration of the request](https://wordpress.stackexchange.com/questions/74676/how-does-object-caching-work), namely autoloaded options/transients, to avoid having to hit the DB for those. So in those cases, clearing the object cache _only effects the WP Cron/Action Scheduler request_. It won't clear cached data for other requests. Is that incorrect?

If so, we can definitely look at disabling it by default. We'd just need to remove the `wp_using_ext_object_cache()` check and only apply the `'action_scheduler_queue_runner_flush_cache'` with the current default `false`. Those changes [were made to accommodate a default batch size of 100](https://github.com/Prospress/action-scheduler/commit/2b7c2a3f12e5984c118acec70ed28c64f903fa6c), which has since been reduced to 25 anyway, so it's not as necessary.

claudiulodro commented:
> Since @claudiulodro is currently specking out lookup tables, that might be a good thing to try and implement this on for populating the tables. I can assist him on that.

Sounds good. Thanks! I should have the first lookup table complete using our existing background processing class by the time AS 2.0 is released. We can then migrate the background processor to the new one.

Commit:
Squashed 'includes/libraries/action-scheduler/' content from commit 6b3282f33

git-subtree-dir: includes/libraries/action-scheduler
git-subtree-split: 6b3282f3324c1896bd7143db9c7c1fdd64765478

Commit:
Merge commit '10c4a515e40b9d3f5903b15e97376bb153e38c61' as 'includes/libraries/action-scheduler'

thenbrent commented:
I've just updated this to PR include [Action Scheduler 2.0.0](https://github.com/Prospress/action-scheduler/releases/tag/2.0.0).

We've just shipped AS 2.0 with Subscriptions 2.3, so it will be widely tested over the coming weeks.

As it's the first major update to AS in 4 years, I expect we'll see a few patch releases in the coming weeks. I'll update either this PR, or the dev branch if this PR is merged before then, as new AS versions are tagged.

claudiulodro commented:
Excellent. Thanks @thenbrent!

kloon commented on includes/class-woocommerce.php:
```diff
@@ -368,6 +368,11 @@ public function includes() {
+		/**
+		 * Libraries
+		 */
+		include_once WC_ABSPATH . 'includes/libraries/action-scheduler/action-scheduler.php';
```
Can we maybe have some sort of check before including since there are other plugins out there making use of this library already.

thenbrent replied:
@kloon it's better to always include this file to make sure if WC has a newer version of AS, that version is loaded. It's also safe to always load that file, because dependency management is built-into Action Scheduler.

Specifically, when a plugin loads the `action-scheduler.php` file, that instance of the Action Scheduler library will register itself and its version with a global `ActionScheduler_Versions` class (via `ActionScheduler_Versions::register()`). Then on `plugins_loaded` with priority `1`, after all plugins have been loaded. Action Scheduler will initialise only the newest version of itself found in the registry.

kloon replied:
OK, thanks for the explanation. Did a bit of further digging and see what you mean.

kloon reviewed (changes requested):
@thenbrent looking good so far and works a treat. Regarding these wc_ functions, the prefix implies to me that it is part of WooCommerce core, is there a way to move these outside the library folder into a seperate file in includes? Something like wc-scheduler-functions.php?

Then is it possible to abstract the functionality a bit? The idea is to give anyone the ability to add their own queue system but still be able to use these functions without it affecting how the queues are processed or scheduled.

thenbrent commented:
> The idea is to give anyone the ability to add their own queue system but still be able to use these functions without it affecting how the queues are processed or scheduled.

Ah good idea @kloon. I'll submit a few extra commits to achieve that abstraction. 👍

Commit:
Merge branch 'master' into issue_18003

Commit:
Add WC_Queue_Interface

To provide an agreed set of public methods that need to be defined
for any class wanting to act as WooCommerce's queue handler.

The require APIs support both delay jobs, by providing APIs to
schedule an action/job/event for some time in the future, and a
non-delayed job API, to add an event to the queue to run as soon
as possible.

Commit:
Add WC_Action_Queue

As a wrapper for Action Scheduler. This implements the WC_Queue_Interface
so it can be used as WC core's job queue. Implementing WC_Queue_Interface
also means this class has more generic and appropriate job queue nomenclature
than Action Scheduler itself. For exsample, it provides a method for
non-delayed jobs, i.e. WC_Queue_Interface::add(), unlike Action Scheduler,
which requires scheduling a job for now(), which while fairly simple, is not
intuitive or immediately obvious to new comers.

This makes it both more clear and convenient to use Action Scheduler
as a simple job queue instead of just a queue for delayed jobs.

Commit:
Add WC_Queue and WC()->queue() wrapper for accessing it

To take care of instantiating a canonical job queue for use across
all of WooCommerce, and by 3rd party code.

Also use this new API for enqueuing webhooks instead of raw Action
Scheduler APIs.

thenbrent commented:
> is it possible to abstract the functionality a bit? The idea is to give anyone the ability to add their own queue system

@kloon I've abstracted Action Scheduler completely now via the commits 9ebfb6f85...e1c7d70b0.

The tl;dr of that changeset is that a new API, `WC()->queue()`, can now be used to add, schedule, cancel or search for queued actions. The object accessed via `WC()->queue()` needs to implement the new `WC_Queue_Interface`, so it is safe to rely on the methods declared in that interface for interacting with the queue.

Apart from making it possible for individual sites or hosts to swap out the queue handler for their own, this abstraction and the introduction of an interface made it possible to overcome some of the UX issues I've seen with Action Scheduler over the years due to its nomenclature, like having no immediately obvious API for adding a job to run ASAP instead of being scheduled for a time in the future - `WC_Queue_Interface::add()` is defined for this purpose, and is used like `WC()->queue()->add()`.

I haven't tested this changeset much yet. I wanted to get the proposed structure in front of you for approval before I spend too long testing it.

If you're happy with it, I'll test it properly. I'll also write up some more docs to include in the [Developer Wiki](http://github.com/woocommerce/woocommerce/wiki) to explain both how to use the new queue APIs, and also the default implementation details, so folks understand why there can be a delay between an action being scheduled and it being processed, especially for large queues, and approaches to deal with that.

> Regarding these wc_ functions, the prefix implies to me that it is part of WooCommerce core

For now, this PR still includes the `wc_` prefix on those functions. However, I've submitted a PR to Action Scheduler core to [change the prefix to `as_`](https://github.com/Prospress/action-scheduler/pull/181). Once that is merged, I'll update this PR to take advantage of that to avoid confusion.

kloon commented:
This is looking great @thenbrent, only issue I can pick is what @om4james mentioned above. Also a question around the cancel method based on the docblock it sounds like it will just remove the next occurance of a job, I would assume for repeating jobs it will cancel everything? If so lets maybe update the docblock to make it clear that no jobs will run again after cancelling.

Regarding the wc_ functions, we would need to be carefull here as to not make it cause conflicts with older version of AS where these functions will still be defined. Perhaps just introduce new functions with different names in core? Perhaps something like wc_*_job since the functionality is now abstracted and you can basically hook any queue system in.

thenbrent commented on includes/queue/class-wc-action-queue.php:
```diff
@@ -0,0 +1,138 @@
+	 *
+	 * @return string The action ID
+	 */
+	public function add( $timestamp, $hook, $args = array(), $group = '' ) {
```
Ah good catch thanks @om4james! I've rebased that error to oblivion. Thanks for looking over the PR! Would love to hear any other feedback you have. :)

thenbrent commented:
> Also a question around the cancel method based on the docblock it sounds like it will just remove the next occurance of a job, I would assume for repeating jobs it will cancel everything? If so lets maybe update the docblock to make it clear that no jobs will run again after cancelling.

Yeah it will. Good catch on the docblock. I've updated it now to clarify.

I originally had that method named `dequeue()` but changed it to `cancel()` as that's the status applied to the action after it's "dequeued" (it's not deleted from the database immediately). I thought cancel was more clear, but let me know if you think `dequeue()` or something else would be more suitable.

> Regarding the wc_ functions, we would need to be carefull here as to not make it cause conflicts with older version of AS where these functions will still be defined.

Because AS only ever loads and instantiates the most recent version of itself, regardless of what other versions are registered, and the [`functions.php` file is loaded in `ActionScheduler::init()`](https://github.com/Prospress/action-scheduler/blob/2.0.0/classes/ActionScheduler.php#L98), I don't expect changing the prefix on those functions to be an issue.

It will result in deprecated notices being thrown for any plugins using the old `wc_` methods still. But there won't be any errors - everything will be scheduled correctly and continue to run.

thenbrent commented:
One thing I'm still concerned about is the first point I mentioned in the potential issues section of [this comment](https://github.com/woocommerce/woocommerce/issues/16971#issuecomment-333656092): 

> By default, Action Scheduler only initiates a new queue at most once per minute. That wouldn't be sufficient frequency for processing orders after checkout. But it's easy to change that with code to run queues via other methods.

That frequency will be fine for webhooks and #18003, but I fear it's a little misleading to provide a new `WC_Action_Queue::add()` / `WC()->queue()->add()` method when the fastest the action can be processed is still potentially that slow. If those methods are used for other things, I'd really want to see an async request like queue handler put in place. That wouldn't be overly hard to implement, and we could add it to either WC or AS core. Just something I want to mention now.

kloon commented:
Picking up a fatal error with the lastest changes `PHP Parse error:  syntax error, unexpected 'get_class' (T_STRING), expecting function (T_FUNCTION) or const (T_CONST) in /srv/www/wc/wp-content/plugins/woocommerce/includes/queue/class-wc-queue.php on line 56`

kloon commented on includes/queue/class-wc-queue.php:
```diff
@@ -0,0 +1,78 @@
+ 	 *
+	 * @return WC_Queue_Interface
+	 */
+	protected validate_instance( $instance ) {
```
Missing function decleration

kloon commented on includes/queue/class-wc-queue.php:
```diff
@@ -0,0 +1,78 @@
+ 	 *
+	 * @return string
+	 */
+	protected static get_class() {
```
Missing function declaration

thenbrent replied:
🤦‍♂️ by "I haven't tested this changeset much yet", what I meant was "at all yet". :)

I've fixed that up and rebased in the fix. :)

claudiulodro commented:
Some notes:
- This works great after fixing the missing function declarations. 
- I love the built-in status screen with all the job statuses. Good feature. 👍 
- This is much simpler to use than [our current background processing library](https://github.com/woocommerce/woocommerce/blob/28a4e88811abd53280cc05f20067ce26e54826e4/includes/libraries/wp-background-process.php) (which requires a new class for each background process we add).
- The queue abstraction looks good. We wouldn't switch the order processing to use the queue until a major release, so that would give us a chance to modify the `add()` methods if needed. Current uses are going to be things that don't need quick response times: webhooks, regenerating images, populating the reports databases.

We can merge this once the missing function declarations are fixed. :)

thenbrent commented:
Thanks @claudiulodro!

> This works great after fixing the missing function declarations.

Fixed and rebased in the fix. :)

> I love the built-in status screen with all the job statuses. Good feature. 👍

It's especially helpful with _Failed_ actions, as sometimes you can even see the error that occurred when attempting to process the action by looking at the logs, e.g. https://cl.ly/420r3U2r0329

> We wouldn't switch the order processing to use the queue until a major release, so that would give us a chance to modify the add() methods if needed. Current uses are going to be things that don't need quick response times: webhooks, regenerating images, populating the reports databases.

Good to know. Sounds like we've got some time before addressing the `add()` method then. I also really like the idea of shipping this just for those less critical use-cases initially, then gradually using it for more critical jobs overtime after it proves itself across WC's broader user base.

> We can merge this once ... the `as_` prefix is used. :)

We've merged the [`as_` prefix PR](https://github.com/Prospress/action-scheduler/pull/181). I want to get one or two more PRs merged into AS core before we tag AS v2.1.0 though. I'll update this late next week once those are merged and we have tagged v2.1.0. :)

franticpsyx commented:
Awesome thread (and PR). Haven't had a chance to dig very deep into AS yet, but it's going to be interesting to see how fast it will replace the current [scheduler](
https://github.com/woocommerce/woocommerce/blob/28a4e88811abd53280cc05f20067ce26e54826e4/includes/libraries/wp-background-process.php).

Just a word of warning -- the current scheduler is used/included by a number of extensions for various tasks, so let's remind ourselves to keep that class file in place :)

kloon commented:
@franticpsyx we will not be removing the current background processing library, if and a big if we do, it will first be deprecated and stay there for many releases before actually being removed.

peterfabian reviewed (commented):
Great work, nice library. Code is easy to read and well structured, which makes it a breeze to follow the logic.

I'm sorry for the nitpicky points sometimes. 

I have a couple of more general questions:
- how do I cancel recurring actions? 
- how to cancel cron-scheduled actions (e.g. if set to run Mon-Fri at 10.00)?

Not sure if 100% needed, but it would feel a bit safer if `save_post_array`, `save_post_schedule` and `save_action_group` ran in a transaction in `ActionScheduler_wpPostStore::save_action` so that either all of them are created or nothing gets created. If I understand it correctly, this is basically the queue::push operation which should be atomic.

I will try to check it a bit more.

claudiulodro commented on includes/libraries/action-scheduler/lib/cron-expression/README.md:
```diff
@@ -0,0 +1,92 @@
+Requirements
+============
+
+- PHP 5.3+
```
It should be OK, as I don't think we'll use CRON format for anything so we won't need to use the parser. I see in this file there is a PHP 5.2 port available if it's a problem.

Commit:
Update WC_Queue_Interface::get_next()

To more closely align return value with other date APIs in WC, namely
WC_Data. Includes changing return value to:

1. a WC_DateTime instead of a timestamp when a scheduled occurrence is found
2. null instead of false, similar to the default of WC_Data::set_date_prop()

thenbrent commented:
Hi @peterfabian. Thanks for the review, I've addressed all of your issues and rebased in the fixes.

Responses to your other items below:

> how do I cancel recurring actions?
> how to cancel cron-scheduled actions (e.g. if set to run Mon-Fri at 10.00)?

The one API cancels all actions: `WC()->queue()->cancel()`.

I've updated the docblock to clarify how this works on recurring actions. Please let me know if the behaviour is still unclear or ambiguous with the updated docblock.

> it would feel a bit safer if save_post_array, save_post_schedule and save_action_group ran in a transaction in ActionScheduler_wpPostStore::save_action so that either all of them are created or nothing gets created. If I understand it correctly, this is basically the queue::push operation which should be atomic.

That's a really good suggestion. I think it's something that should be opened as an Issue on [the Action Scheduler](https://github.com/prospress/action-scheduler) repository though to address in the core library as it affects all implementations of it.

thenbrent commented:
I've just added a new commit to this to bring the return value of `WC_Queue_Interface::get_next()` inline with _the WC Way_:tm: rather than the Action Scheduler way.

The return value is now:

1. a `WC_DateTime` instead of a timestamp when a scheduled occurrence is found
1. `null` instead of `false`, when no next scheduled action is found, similar to the default of `WC_Data::set_date_prop()`

I could do similar with the `WC_Queue_Interface` schedule methods to make the first parameter a `WC_DateTime` instead of a timestamp. Let me know if you think that makes sense.

kloon reviewed (approved):
This looks good to merge now, I can take care of outstanding PHPCS issues once merged and then we can take it from there to implement the outstanding changes that will form part of AS 2.1 when that gets released.
